src/backend/optimizer/path/allpaths.c

```diff
@@ -2125,15 +2125,33 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			cteplaninfo->subplan = (subplan->flow->locustype != CdbLocusType_SegmentGeneral) ?
+			                       prepare_plan_for_sharing(cteroot, subplan) :
+			                       subplan;
+
 			cteplaninfo->subroot = subroot;
 		}
 
 		/*
 		 * Create another ShareInputScan to reference the already-created
-		 * subplan.
+		 * subplan if not avoiding sharing. Sharing SegmentGeneral subplan may
+		 * lead to deadlock when executed with 1-gang and joined with Hashed.
 		 */
-		subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
+		if (cteplaninfo->subplan->flow->locustype != CdbLocusType_SegmentGeneral)
+		{
+			subplan = share_prepared_plan(cteroot, cteplaninfo->subplan);
+		}
+		else
+		{
+			/*
+			 * If we are not sharing and subplan was created just now, use it.
+			 * Otherwise, make a copy of it to avoid construction of DAG
+			 * instead of a tree.
+			 */
+			if (subplan == NULL)
+				subplan = (Plan *) copyObject(cteplaninfo->subplan);
+		}
+
 		subroot = cteplaninfo->subroot;
```

Comment:
Is it safe to use the same subroot? Seems like yes, but just to make sure.

Reply:
Seems that it does not affect anything.